Orange/data/table.py

```diff
@@ -33,146 +28,176 @@ def get_sample_datasets_dir():
-                if self.sparse_y:
-                    self.table._Y[self.row_index, key - len(self._x)] = value
-        else:
-            self._metas[-1 - key] = value
-            if self.sparse_metas:
-                self.table.metas[self.row_index, -1 - key] = value
-
-    def _str(self, limit):
-        def sp_values(matrix, variables):
-            if not sp.issparse(matrix):
-                return Instance.str_values(matrix[row], variables, limit)
-            begptr, endptr = matrix.indptr[row:row + 2]
-            rendptr = endptr if not limit else min(endptr, begptr + 5)
-            variables = [variables[var]
-                         for var in matrix.indices[begptr:rendptr]]
-            s = ", ".join(
-                "{}={}".format(var.name, var.str_val(val))
-                for var, val in zip(variables, matrix.data[begptr:rendptr]))
-            if limit and rendptr != endptr:
-                s += ", ..."
-            return s
-
-        table = self.table
-        domain = table.domain
-        row = self.row_index
-        s = "[" + sp_values(table.X, domain.attributes)
-        if domain.class_vars:
-            s += " | " + sp_values(table._Y, domain.class_vars)
-        s += "]"
-        if self._domain.metas:
-            s += " {" + sp_values(table.metas, domain.metas) + "}"
-        return s
 
-    def __str__(self):
-        return self._str(False)
+# noinspection PyPep8Naming
+class Table(Storage, DataFrame):
+    _WEIGHTS_COLUMN = ContinuousVariable("__weights__")
+    _WEIGHTS_COLUMN.is_weight = True
 
-    def __repr__(self):
-        return self._str(True)
+    # a counter for indexing rows, important for deterministically selecting rows
+    # and keeping pandas indices sane
+    _next_instance_id = 0
+    _next_instance_lock = Lock()
 
+    __file__ = None
 
-class Columns:
-    def __init__(self, domain):
-        for v in chain(domain, domain.metas):
-            setattr(self, v.name.replace(" ", "_"), v)
+    # custom properties, preserved through pandas manipulations
+    _metadata = ['name', '__file__', '_columns_X', '_columns_Y', '_columns_meta']
 
+    @property
+    def _constructor(self):
+        """Proper pandas extension as per http://pandas.pydata.org/pandas-docs/stable/internals.html"""
+        return Table
 
-# noinspection PyPep8Naming
-class Table(MutableSequence, Storage):
-    __file__ = None
-    name = "untitled"
+    @property
+    def _constructor_sliced(self):
+        """Proper pandas extension as per http://pandas.pydata.org/pandas-docs/stable/internals.html"""
+        return TableSeries
+
+    @property
+    def _constructor_expanddim(self):
+        return TablePanel
+
+    @property
+    def domain(self):
+        # TODO: probably return a Domain object constructed from columns
+        return self.columns
+
+    @property
+    def columns_X(self):
+        """A read-only list of X column variables."""
+        return [c for c in self.columns if c in self._columns_X]
```

Comment:
I think we'll use a variant of the first approach. Getting the columns through Domain (@astaric and I had a talk today about Domain, I'll fill you in tomorrow), then selecting the data through those columns.